src/Common/src/System/Data/ProviderBase/DbConnectionPoolGroup.cs

```diff
@@ -164,32 +165,30 @@ internal DbConnectionPool GetConnectionPool(DbConnectionFactory connectionFactor
                 {
                     if (!_poolCollection.TryGetValue(currentIdentity, out pool)) // find the pool
                     {
+                        // this may result in more pools than are used being created but they will be shutdown if they are not used
+                        // and the lock below is a hot point of contention so we need to do as little as possible inside it
+                        DbConnectionPoolProviderInfo connectionPoolProviderInfo = connectionFactory.CreateConnectionPoolProviderInfo(this.ConnectionOptions);
```

Comment:
I had a quick look, and realized that this piece of code was moved inside the lock in the commit at https://github.com/dotnet/corefx/commit/4839aaf6dbbf2d453b9c27283325e0df90db5f5d

https://github.com/dotnet/corefx/pull/23722

Does the check below, guarantee that the unnecessary pools will be shutdown? Is that why it is OK to move the pool creation outside the lock?

Reply:
It is the intention that multiple pools could be created but not started. It's a little wasteful but not terrible. The only way to prevent that new pool being shutdown is to add it to the concurrent dictionary successfully. It doesn't so much differently I just wanted to minimize the time under the lock.

Reply:
Yeah, I think the approach is fine. Just wanted to make sure that your change don't uncover the bug that #23722 was fixing. Seems like it won't.